radix-transaction-scenarios/src/executor.rs

```diff
@@ -0,0 +1,592 @@
+    starting_nonce: u32,
+    /// How the executor should handle nonces and how it should get the next nonce.
+    next_scenario_nonce_handling: ScenarioStartNonceHandling,
+    /// The network definition to use in the execution of the scenarios.
+    network_definition: NetworkDefinition,
+
+    /* Callbacks */
+    /// A callback that is called when a new protocol requirement is encountered. This can be useful
+    /// for clients who wish to apply protocol updates they wish.
+    on_new_protocol_requirement_encountered: F1,
+    /// A callback that is called when a scenario transaction is executed.
+    on_transaction_executed: F2,
+    /// A callback that is called when a new scenario is started.
+    on_scenario_start: F3,
+    /// A callback that is called after bootstrapping if bootstrapping is enabled.
+    after_bootstrap: F4,
+}
+
+pub type DefaultTransactionScenarioExecutor<D> = TransactionScenarioExecutor<
+    D,
+    DefaultWasmEngine,
+    NoExtension,
+    fn(&NetworkDefinition, ProtocolUpdate, &mut D),
+    fn(&ScenarioMetadata, &NextTransaction, &TransactionReceiptV1, &D),
+    fn(&ScenarioMetadata),
+    fn(&NetworkDefinition, &mut D),
+>;
+
+impl<D, W, E, F1, F2, F3, F4> TransactionScenarioExecutor<D, W, E, F1, F2, F3, F4>
+where
+    D: SubstateDatabase + CommittableSubstateDatabase,
+    W: WasmEngine,
+    E: NativeVmExtension,
+    F1: FnMut(&NetworkDefinition, ProtocolUpdate, &mut D),
+    F2: FnMut(&ScenarioMetadata, &NextTransaction, &TransactionReceiptV1, &D),
+    F3: FnMut(&ScenarioMetadata),
+    F4: FnMut(&NetworkDefinition, &mut D),
+{
+    pub fn new(
+        database: D,
+        network_definition: NetworkDefinition,
+    ) -> DefaultTransactionScenarioExecutor<D> {
+        DefaultTransactionScenarioExecutor::<D> {
+            /* Environment */
+            database,
+            scrypto_vm: ScryptoVm::default(),
+            native_vm_extension: NoExtension,
+            /* Execution */
+            filter: None,
+            registered_scenarios: Default::default(),
+            bootstrap: true,
+            starting_nonce: 0,
+            next_scenario_nonce_handling:
+                ScenarioStartNonceHandling::PreviousScenarioEndNoncePlusOne,
+            network_definition,
+            /* Callbacks */
+            on_new_protocol_requirement_encountered: |network_definition, protocol_update, db| {
+                protocol_update
+                    .generate_state_updates(db, network_definition)
+                    .into_iter()
+                    .for_each(|state_updates| {
+                        db.commit(&state_updates.create_database_updates::<SpreadPrefixKeyMapper>())
+                    })
+            },
+            on_transaction_executed: |_, _, _, _| {},
+            on_scenario_start: |_| {},
+            after_bootstrap: |_, _| {},
+        }
+        // Genesis Scenarios.
+        .register_scenario_with_create_fn::<TransferXrdScenarioCreator>()
```

Comment:
We filter the list of all scenarios in the node based on their name in their metadata, so this breaks node compatibility.

Reply:
You can still filter them by name using `ScenarioFilter::ExactScenarios`, the registered scenarios is esentially the list of all scenarios, does this fit the node's requirements?

Reply:
No it doesn't. As mentioned, we can't use this executor on the node -- because we need to be able to commit ourselves.

We do:
```
        let mut scenario = self
            .find_scenario(epoch, nonce, scenario_name)
            .unwrap_or_else(|| {
                panic!(
                    "Could not find scenario with logical name: {}",
                    scenario_name
                )
            });
```

Where:
```
    fn find_scenario(
        &self,
        epoch: Epoch,
        next_nonce: u32,
        scenario_name: &str,
    ) -> Option<Box<dyn ScenarioInstance>> {
        for scenario_builder in get_builder_for_every_scenario() {
            let scenario =
                scenario_builder(ScenarioCore::new(self.network.clone(), epoch, next_nonce));
            if scenario.metadata().logical_name == scenario_name {
                return Some(scenario);
            }
        }
        None
    }
```

Reply:
(btw - for the node-complexity with transaction execution, see `prepare_scenario_transaction` for the gnarly details :P - the way we execute and then commit scenario transactions is just very different to the simple world in scrypto 👍 )